src/libmongoc/tests/unified/entity-map.c

```diff
@@ -269,71 +271,118 @@ command_succeeded (const mongoc_apm_command_succeeded_t *succeeded)
+                                   bsonAs (cstr)))),
+            // An unknown event name is a hard-error:
+            else(do(test_error ("Unknown event type '%s'", bsonAs (cstr))))))),
+      // Command events to ignore
+      find (
+         key ("ignoreCommandMonitoringEvents"),
+         if (not(type (array)),
+             then (error ("'ignoreCommandMonitoringEvents' must be an array"))),
+         visitEach (if (not(type (utf8)),
+                        then (error ("Every 'ignoreCommandMonitoringEvents' "
+                                     "element must be a string")))),
+         storeDocDupPtr (entity->ignore_command_monitoring_events)),
+      // Parse the serverApi, if present
+      find (
+         key ("serverApi"),
+         if (not(type (doc)), then (error ("'serverApi' must be a document"))),
+         parse ( // The "version" string is required first:
+            find (keyWithType ("version", utf8), do({
+                     mongoc_server_api_version_t ver;
+                     if (!mongoc_server_api_version_from_string (bsonAs (cstr),
+                                                                 &ver)) {
+                        bsonParseError = "Invalid serverApi.version string";
+                     } else {
+                        api = mongoc_server_api_new (ver);
+                     }
+                  })),
+            else(error ("Missing 'version' property in 'serverApi' object")),
+            // Toggle strictness:
+            find (key ("strict"),
+                  if (not(type (bool)),
+                      then (error ("'serverApi.strict' must be a bool"))),
+                  do(mongoc_server_api_strict (api, bsonAs (bool)))),
+            // Toggle deprecation errors:
+            find (
+               key ("deprecationErrors"),
+               if (not(type (bool)),
+                   then (error ("serverApi.deprecationErrors must be a bool"))),
+               do(mongoc_server_api_deprecation_errors (api, bsonAs (bool)))))),
+      // Toggle observation of sensitive commands
+      find (key ("observeSensitiveCommands"),
+            if (not(type (bool)),
+                then (error ("'observeSensitiveCommands' must be a bool"))),
+            do({
+               bool *p = entity->observe_sensitive_commands =
+                  BSON_ALIGNED_ALLOC0 (bool);
+               *p = bsonAs (bool);
+            })),
+      // Which events should be available as entities:
+      find (key ("storeEventsAsEntities"),
+            if (not(type (array)),
+                then (error ("'storeEventsAsEntities' must be an array"))),
+            visitEach (if (not(type (utf8)),
+                           then (error ("Every 'storeEventsAsEntities' element "
+                                        "must be a string")))),
+            /* TODO: CDRIVER-3867 Comprehensive Atlas Testing */
+            error ("'storeEventsAsEntities' is not yet supported")));
```

Comment:
I was unable to formulate a way for `parse` to do this cleanly without requiring arbitrary allocations. `parse` doesn't keep track of which elements it has inspected. Because a document can have arbitrarily many elements, one would need a multiset to keep track of them. The `find(not(...))` pattern will work, even if it is less pretty.

Something similar could be done by using `case()` with the visit API instead:

```c
bsonVisitEach(doc,
  case(when(key("foo"), ...),
       when(key("bar"), ...),
       when(key("baz"), ...),
       else(error("Unrecognized key"))));
```

But: This form doesn't support a "required elements" semantics, which would have to be implemented manually.

Reply:
I was able to implement this in an unlikely-to-allocate manner without any key tracking required. `parse` now keeps an array of bits, with each bit corresponding to an element. When `find` or `require` visits an element, it sets a one bit in the array for the element. The `visitOthers` op iterates over the document again and applies a visitor to every element that has a zero bit. This will only allocate for very large documents that require many bits. (I also simplified the implementation of `find` and `require` significantly.)